### PR TITLE
LSP: resolve 'go-to-definition' for calls, including in generic instantiations

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR implements go-to-definition for calls. It does so by introducing a new segment type (call segments) to denote things called at a given position.

gotodefcall

## Segment Lists Overhaul
This PR also implements a major overhaul of how `chpl-language-server`'s segment lists work. Specifically, it makes the segment lists work when items stored therein have arbitrary overlapping behavior (nesting, partial overlaps, disjointedness). Consider the following representation of a list of things which are found in a source file (e.g., names, calls, etc.)

```
        |------------| A
               |-------------| B
                       |--| C
                                  |---------| D
```

Prior to this PR, CLS stored a sorted list of these items, and used a binary search across the elements' start positions to find the item under the cursor. However, this didn't work in situations such as the B-C overlap in the diagram above. A binary search when the cursor is past C would find C as the next visible start position, and, since that position doesn't contain the cursor, return nothing. However, it should return B, which resumes after C is over. In other words, overlapping segments would "interrupt" ranges in CLS.

This is going to be a problem for nested instantiations (though they are currently not possible due to limitations of Dyno's resolver), since the location of the "inner" generic function overlaps with the location of the "outer" generic function, and thus, the region after the "inner" function will not be correctly determined to belong to the "outer" function (it would be interrupted). Since go-to-called-function support relies on instantiations, I elected to improve the situation in this PR.

I did so by additionally "flattening" the representation of item ranges, so that no parts of it overlap. The flattened representation of the above segment list is:

```
        |------|-------|--|--|----|---------|
         A      B       C  B  None D

```

Binary searches are performed using start positions of the flattened segments (the labeled `|`s in the above diagram). 

## Making Use of Instantiation Segments
To construct call segments (and thus support go-to-called-function), this PR makes use of instantiation segments. Specifically, when an instantiation at a particular range changes, CLS will iterate the affected instnatiation segments (which correspond to regions-within-an-instantiation in the user buffer) and re-run call resolution within, given the instantiation information.

Reviewed by @jabraham17 -- thanks!

## Testing
- [x] python tests for CLS

### CODE REVIEW

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 1]
```diff
@@ -844,18 +1026,22 @@ def _search_instantiations(
...
+            # functions. There may be multiple instantiations per file,
+            # and until the user has selected one, we shouldn't enable go-to-def
+            # on calls within then.
+            include_in_segments = via is None
+            if include_in_segments:
+                self.call_segments.append(
+                    ResolvedPair(
+                        NodeAndRange(node.called_expression()), NodeAndRange(fn)
+                    )
+                )
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1044]: This is the feature I was talking about the other day. Being able to "goto-defintion" on a generic function and have the editor offer the user a list of possible locations is a useful feature. I think you should open an issue for a feature request for this

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 2]
```diff
@@ -869,6 +1055,90 @@ def _search_instantiations(
 
             self._search_instantiations(fn, via=sig)
 
+    def find_decl_by_unique_id(self, unique_id) -> Optional[NodeAndRange]:
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1058]: nit, type annotation for `unique_id`

Comment (jabraham17) [reply tools/chpl-language-server/src/chpl-language-server.py:1058]: same for other functions

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 3]
```diff
@@ -869,6 +1055,90 @@ def _search_instantiations(
...
+            if elt_idx in segments_for_elt:
+                return segments_for_elt[elt_idx]
+
+            inst, sig = value_at_segment
+            calls_in_inst = PositionList[ResolvedPair](lambda x: x.ident.rng)
+
+            for node in chapel.preorder(inst.node):
+                if not isinstance(node, chapel.FnCall):
+                    continue
+
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1104]: why not  `chapel.each_matching(inst.node, chapel.FnCall, iterator=chapel.postorder`?

File: tools/chpl-language-server/test/basic_resolver.py [hunk 4]
```diff
@@ -54,6 +54,68 @@ async def test_go_to_record_def(client: LanguageClient):
         await check_goto_type_def(client, doc, pos((3, 9)), pos((0, 7)))
 
 
+@pytest.mark.asyncio
+async def test_go_to_call_basic(client: LanguageClient):
+    """
+    Ensure that 'go to definition' on a type actually works.
```
Comment (jabraham17) [inline tools/chpl-language-server/test/basic_resolver.py:60]: nit, comment is wrong

note that the comment should be updated for all the functions

File: frontend/lib/parsing/ParserContextImpl.h [hunk 5]
```diff
@@ -1489,8 +1489,8 @@ CommentsAndStmt ParserContext::buildFunctionDecl(YYLTYPE location,
   bool primaryMethod = false;
   if (currentScopeIsAggregate()) {
     if (fp.receiver == nullptr) {
-      fp.receiver = buildThisFormal(location, fp.thisIntentLoc, fp.thisIntent,
-                                    nullptr, nullptr);
+      fp.receiver = buildThisFormal(fp.thisIntentLoc, fp.thisIntentLoc,
+                                    fp.thisIntent, nullptr, nullptr);
       primaryMethod = true;
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply frontend/lib/parsing/ParserContextImpl.h:1494]: > In this case, the location of this is ....something? Probably just an invalid location, or maybe bison does something weird?

In this case the location is the empty space right before `foo`, which I've found to be quite nice in testing.

If you think about it, in the `ref` case, you can think of `ref` as a component of `ref this`, where the rest is invisible, so it makes _some_ sense to me. In any case, I really can't imagine a better source location to associate with it.

File: tools/chpl-language-server/src/chpl-language-server.py [hunk 6]
```diff
@@ -1170,7 +1168,8 @@ def rebuild_index(self):
             # TODO: suppress resolution errors due to false-positives
             # this should be removed once the resolver is finished
             with self.context.context.track_errors() as _:
-                self._search_instantiations(asts)
+                for ast in asts:
+                    self._search_instantiations(ast)
```
Comment (jabraham17) [inline tools/chpl-language-server/src/chpl-language-server.py:1172]: why is this change needed? doesn't `chapel.each_matching` handle that? it should only yield things of the pattern

Comment (DanilaFe) [reply tools/chpl-language-server/src/chpl-language-server.py:1172]: It's not needed, but I prefer it this way rather than relying on the fact that both types yield the same thing.

**Review by jabraham17 — APPROVED** [review]
